Docs: Document fine-grained access control

vtorosyan opened the pull request:
**What this PR does / why we need it**:

The PR introduces new docs for the fine-grained access control and updates relevant connected pages. The [issue](https://github.com/grafana/grafana-enterprise/issues/1282) describes the scope of this PR. [This comment](https://github.com/grafana/grafana/pull/33563#issuecomment-833501418) contains few screenshots.

**Note:** If you know any page that needs to be updated or a new documentation to be added, please update the issue so that we can keep track of changes.

**Big Question**

At the moment, I'm using "Access Control" name for the new feature. However, I consider Fine-grained access control as a better alternative. Main reasoning behind it is that "access control" is a clearly overused wording and it is very hard to distinguish which access control we mean when talking about it in the new docs. While I see us eventually moving towards having one unified Access Control, thus naming the capability like that, now we are far away from that.

We have also covered this a bit in the [Slack conversation](https://raintank-corp.slack.com/archives/C01LU2Y5527/p1620295553224800).

**UPDATE:** I have decided to switch the term to `Fine-grained access control`, for the above mentioned reasons. Please do let me know what you think about it and if there are any concerns or suggestions. Thanks~

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana-enterprise/issues/1282

vtorosyan committed:
Access control documentation

vtorosyan committed:
Merge branch 'master' into vt/ac-docs

vtorosyan committed:
Adding docs for the API

vtorosyan committed:
Few improvements

vtorosyan committed:
More docs

vtorosyan committed:
Start admin docs

vtorosyan committed:
Merge branch 'master' into vt/ac-docs

vtorosyan committed:
Adding admin API docs

vtorosyan committed:
Reorder and more docs

vtorosyan committed:
Update user API and reporting API docs

vtorosyan committed:
Update reporting docs

gamab committed:
Correct little typo

vtorosyan committed:
Update org roles and permissions pages

gamab commented on docs/sources/enterprise/access-control/concepts/roles.md:
```diff
@@ -0,0 +1,58 @@
+
+# Built-in role assignments
+
+To control what your users can access or not, you can assign or unassign a role to the existing Organization roles, or to a Grafana Admin. These assignments are called built-in role assignments.
```
```suggestion
To control what your users can access or not, you can assign or revoke a role to the existing Organization roles, or to a Grafana Admin. These assignments are called built-in role assignments.
```

gamab commented on docs/sources/http_api/access_control.md:
```diff
@@ -0,0 +1,536 @@
+
+Code | Description
+--- | --- | 
+200 | Role was unassigned from built-in role.
```
```suggestion
200 | Role was revoked from built-in role.
```

vtorosyan committed:
Merge branch 'vt/ac-docs' of github.com:grafana/grafana into vt/ac-docs

vtorosyan committed:
How to guides

vtorosyan commented:
Adding couple of screenshots to give an idea about how changes look like.

**Nav menu**
<img width="312" alt="Screenshot 2021-05-06 at 14 55 38" src="https://user-images.githubusercontent.com/1861190/117301884-5d08cb80-ae7b-11eb-81a1-f731fdbb462d.png">

**Roles page with all predefined roles**
<img width="1177" alt="Screenshot 2021-05-06 at 14 55 52" src="https://user-images.githubusercontent.com/1861190/117301930-67c36080-ae7b-11eb-898a-8e2a0325aaf1.png">

**Permissions**
<img width="1164" alt="Screenshot 2021-05-06 at 14 56 07" src="https://user-images.githubusercontent.com/1861190/117301970-70b43200-ae7b-11eb-85aa-598c2175c130.png">

**HTTP API parts**
<img width="1003" alt="Screenshot 2021-05-06 at 14 56 27" src="https://user-images.githubusercontent.com/1861190/117302002-7ad63080-ae7b-11eb-9270-5c8d1f91106a.png">

**Banner with note**
<img width="1199" alt="Screenshot 2021-05-06 at 14 56 33" src="https://user-images.githubusercontent.com/1861190/117302031-83c70200-ae7b-11eb-9f66-79bc3336dbb7.png">

gamab committed:
Started work on provisioning

vtorosyan committed:
Adding a link to provisioning

vtorosyan committed:
Merge branch 'master' into vt/ac-docs

gamab committed:
Add a supported settings section

gamab committed:
Listing the current validation rules

gamab committed:
Add missing required to the example

vtorosyan committed:
Merge branch 'main' into vt/ac-docs

vtorosyan committed:
Refactoring and fixes

vtorosyan commented:
I am opening this up for review. While there is a lot to improve, I believe it's in a good place to start getting feedback.

sakjur reviewed (commented):
🎉 This is great work!

I left a bunch of comments where the most important ones are places where I've found the documentation to be slightly inaccurate/incomplete.

oddlittlebird reviewed (dismissed):
Wow, lots of great content in this PR! When @osg-grafana is back tomorrow, let's discuss how to make it even better.

mjseaman committed:
grammar update

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

mjseaman commented on docs/sources/enterprise/access-control/_index.md:
```diff
@@ -0,0 +1,28 @@
+
+> Fine-grained access control is available behind the `accesscontrol` feature flag in Grafana Enterprise 8.0+.
+
+Fine-grained access control provides a standardized way of granting, changing and revoking access to Grafana resources. 
```
Agree with @vtorosyan's sentiment. Perhaps where we use the term fine-grained access control other than the title and definition, we can find more specific terms like "custom roles" or "updating permissions for a role."

mjseaman committed:
grammar

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>
Co-authored-by: Emil Tullstedt <emil.tullstedt@grafana.com>

mjseaman committed:
grammar and wording updates

vtorosyan committed:
Fix the ref links breaking the build

vtorosyan committed:
Merge branch 'main' into vt/ac-docs

vtorosyan commented on docs/sources/enterprise/access-control/_index.md:
```diff
@@ -0,0 +1,28 @@
+weight = 100
++++
+
+# Fine-grained access control
```
This is a bit tricky to do, as essentially everyone who has required permissions can manage access control. That said, when Grafana starts, in the beginning only Grafana Server Admin can do that, and Grafana Server Admin can later decide that someone else needs also to be able to do it and give them permissions.

I will look into where it would make sense to mention this somehow.

vtorosyan commented on docs/sources/enterprise/access-control/provisioning.md:
```diff
@@ -0,0 +1,103 @@
+### Built-in role assignments
+
+1. `name` must be one of the Organization roles (`Viewer`, `Editor`, `Admin`) or `Grafana Admin`. 
+1. When `orgId` is not specified, it inherits the `orgId` from `role`.
```
Not sure, this was from Gabriel from the provisioning work. I will try to test.

vtorosyan commented on docs/sources/enterprise/access-control/concepts/permissions.md:
```diff
@@ -0,0 +1,74 @@
+weight = 130
++++
+
+# Permissions 
```
It is not applicable at all to OSS Grafana from usage and product perspective. That said, a lot of code is in the OSS Grafana, so it might be valuable to cover some parts of this in a developer guide. That would be out of scope of this PR though.

vtorosyan commented on docs/sources/enterprise/access-control/concepts/permissions.md:
```diff
@@ -0,0 +1,74 @@
+To grant permissions to your users, you can create [built-in role assignments]({{< relref "./roles.md#built-in-role-assignments" >}}).
+
+A permission is defined by an `action` and a `scope`. 
+When evaluating a decision for access, user will be allowed to perform a specific action on a resource if they have a permission with relevant `action` and `scope`.
```
They don't have to be, I'll remove.

vtorosyan commented on docs/sources/enterprise/access-control/concepts/permissions.md:
```diff
@@ -0,0 +1,74 @@
+`scope` describes where ant action can be performed. For example, you can grant a user access to read a specific user profile, by associating a permission with scope `users:<userId>` to the relevant role.
+You can also combine multiple scopes if necessary, by using `/` as a delimiter. Refer to [Managing roles and permissions]({{< relref "../managing-roles-permissions.md" >}}) to learn more about which permissions are used for which resources.
+
+## Available permissions 
```
Just to make sure that I understand right, you mean the heading should be H1 not H2? 🤔

vtorosyan commented on docs/sources/enterprise/access-control/concepts/roles.md:
```diff
@@ -0,0 +1,62 @@
+
+Refer to the [Access Control API]({{< relref "../../../http_api/access_control.md" >}}) to list available roles with permissions.
+
+## Global and organization roles
```
I did not want to put in Scope, as we have a scope concept for permission, and this one is different. I understand the confusion, I will think how to improve this.

vtorosyan commented on docs/sources/enterprise/access-control/_index.md:
```diff
@@ -0,0 +1,28 @@
+Fine-grained access control provides a standardized way of granting, changing and revoking access to Grafana resources. 
+The fine-grained access control works hand-in-hand with the existing [Grafana permissions]({{< relref "../../permissions/_index.md" >}}) and allows you to control actions users can perform in a granular way.
+
+To learn more about how fine-grained access control works, refer to [Concepts]({{< relref "./concepts/_index.md" >}}) and to [Manage roles and permissions]({{< relref "./managing-roles-permissions.md" >}}).
```
Honestly the easiest "fix" would be to remove this hierarchy and have flat structure, so roles, permissions and custom roles will be in the same level as the index 🤔

vtorosyan commented on docs/sources/enterprise/access-control/how-to-examples.md:
```diff
@@ -0,0 +1,20 @@
+weight = 130
++++
+
+# How-to examples
```
Yes it's a todo to add more examples.

vtorosyan committed:
Refactor docs

vtorosyan committed:
More comments handled

vtorosyan commented:
Thanks everyone for contributions and suggestions. Great feedback!
I've tried to touch all raised concerns and unfortunately this means a major refactoring. I'd recommend to review from start, rather then new commits only.

Major changes:
- Structure is flattened. I have removed "Concepts" subsection, "Managing roles and permissions" section and simplified things a lot.
- Tried to follow https://github.com/grafana/grafana/tree/main/contribute/templates#documentation-templates whenever possible

vtorosyan committed:
Merge branch 'main' into vt/ac-docs

osg-grafana commented on docs/sources/enterprise/access-control/_index.md:
```diff
@@ -0,0 +1,42 @@
++++
+title = "Fine-grained access control"
+description = "Understand how to grant, change or revoke access to the Grafana resources"
```
```suggestion
description = "Grant, change, or revoke access to Grafana resources"
```

osg-grafana commented on docs/sources/enterprise/access-control/_index.md:
```diff
@@ -0,0 +1,42 @@
+
+# Fine-grained access control
+
+> **Note:** Fine-grained access control is currently in beta. Expect changes in future releases.
```
```suggestion
> **Note:** Fine-grained access control is in beta, and you can expect changes in future releases.
```

vtorosyan replied:
One day, would be nice to update these in other docs. Since I suspect many people, including me are just looking at the previous examples and copy pasting it.

osg-grafana commented on docs/sources/enterprise/access-control/_index.md:
```diff
@@ -0,0 +1,42 @@
+
+> **Note:** Fine-grained access control is currently in beta. Expect changes in future releases.
+
+Fine-grained access control provides a standardized way of granting, changing, and revoking access to view and modify Grafana resources, like users and reports. 
```
```suggestion
Fine-grained access control provides a standardized way of granting, changing, and revoking access when it comes to viewing and modifying Grafana resources, such as users and reports. 
```

osg-grafana commented on docs/sources/enterprise/access-control/_index.md:
```diff
@@ -0,0 +1,42 @@
+> **Note:** Fine-grained access control is currently in beta. Expect changes in future releases.
+
+Fine-grained access control provides a standardized way of granting, changing, and revoking access to view and modify Grafana resources, like users and reports. 
+Fine-grained access control works hand-in-hand with the existing [Grafana permissions]({{< relref "../../permissions/_index.md" >}}) and allows you to control actions users can perform in a more granular way.
```
```suggestion
Fine-grained access control works with [Grafana permissions]({{< relref "../../permissions/_index.md" >}}), and it allows you granular control of users’ actions.
```

osg-grafana commented on docs/sources/enterprise/access-control/_index.md:
```diff
@@ -0,0 +1,42 @@
+Fine-grained access control works hand-in-hand with the existing [Grafana permissions]({{< relref "../../permissions/_index.md" >}}) and allows you to control actions users can perform in a more granular way.
+
+To learn more about how fine-grained access control works, refer to [Roles]({{< relref "./roles.md" >}}) and [Permissions]({{< relref "./permissions.md" >}}).
+Refer to [fine-grained access control usage scenarios]({{< relref "./usage-scenarios.md" >}}) to see step-by-step instructions and understand how you can use the fine-grained access control system.
```
```suggestion
To use the fine-grained access control system, refer to [Fine-grained access control usage scenarios]({{< relref "./usage-scenarios.md" >}}).
```

osg-grafana commented on docs/sources/enterprise/access-control/_index.md:
```diff
@@ -0,0 +1,42 @@
+## Enable fine-grained access control
+
+Fine-grained access control is available behind the `accesscontrol` [feature toggle]({{< relref "../../administration/configuration.md#feature_toggles" >}}) in Grafana Enterprise 8.0+.
+Refer to [Configuration]({{< relref "../../administration/configuration.md" >}}) to understand how you can enable it in different environments.
```
```suggestion
To enable it in different environments, refer to [Configuration]({{< relref "../../administration/configuration.md" >}}).
```

osg-grafana commented on docs/sources/enterprise/access-control/_index.md:
```diff
@@ -0,0 +1,42 @@
+
+## Access management
+
+When making a decision about access, the fine-grained access control takes into the account the following main parts:
+- _who_ has an access (`identity`).
+- _what they can do_ and on which _Grafana resource_ (`role`).
```
```suggestion
Fine-grained access control considers a) _who_ has an access (`identity`), and b) _what they can do_ and on which _Grafana resource_ (`role`).
```

osg-grafana commented on docs/sources/enterprise/access-control/_index.md:
```diff
@@ -0,0 +1,42 @@
+- _who_ has an access (`identity`).
+- _what they can do_ and on which _Grafana resource_ (`role`).
+
+You can grant, change, or revoke access to _users_ (`identity`). When an authenticated user tries to access a Grafana resource, the authorization system checks the required [fine-grained permissions]({{< relref "./permissions.md" >}}) for the resource and determines whether the action is allowed or not. 
```
```suggestion
You can grant, change, or revoke access to _users_ (`identity`). When an authenticated user tries to access a Grafana resource, the authorization system checks the required fine-grained permissions for the resource and determines whether or not the action is allowed. Refer to [Fine-grained permissions]({{< relref "./permissions.md" >}}).
```

osg-grafana commented on docs/sources/enterprise/access-control/_index.md:
```diff
@@ -0,0 +1,42 @@
+
+You can grant, change, or revoke access to _users_ (`identity`). When an authenticated user tries to access a Grafana resource, the authorization system checks the required [fine-grained permissions]({{< relref "./permissions.md" >}}) for the resource and determines whether the action is allowed or not. 
+
+To grant or revoke access to your users, you can create or remove a [built-in role assignments]({{< relref "./roles.md#built-in-role-assignments" >}}).
```
```suggestion
To grant or revoke access to your users, create or remove built-in role assignments. For more information, refer to [Built-in role assignments]({{< relref "./roles.md#built-in-role-assignments" >}}).
```

osg-grafana replied:
```suggestion
To grant or revoke access to your users, create or remove built-in role assignments. For more information, refer to [Built-in role assignments]({{< relref "./roles.md#built-in-role-assignments" >}}).
```

osg-grafana commented on docs/sources/enterprise/access-control/_index.md:
```diff
@@ -0,0 +1,42 @@
+
+## Resources with fine-grained permissions
+
+Fine-grained access control is in beta and [permissions]({{< relref "./permissions.md" >}}) are available only for a subset of resources.
```
```suggestion
Fine-grained access control permissions are available for a subset of resources. For more information, refer to [Permissions]({{< relref "./permissions.md" >}}).
```

osg-grafana commented on docs/sources/enterprise/access-control/_index.md:
```diff
@@ -0,0 +1,42 @@
+## Resources with fine-grained permissions
+
+Fine-grained access control is in beta and [permissions]({{< relref "./permissions.md" >}}) are available only for a subset of resources.
+Refer to the relevant API guide from below list to learn more about specific endpoints where you can use access control.
+
+- [Access Control API]({{< relref "../../http_api/access_control.md" >}})
+- [Admin API]({{< relref "../../http_api/admin.md" >}})
+- [Organization API]({{< relref "../../http_api/org.md" >}})
+- [Reporting API]({{< relref "../../http_api/reporting.md" >}})
+- [User API]({{< relref "../../http_api/user.md" >}})
```
```suggestion
To learn more about specific endpoints where you can use access control, refer to the relevant API guide:

- [Access Control API]({{< relref "../../http_api/access_control.md" >}})
- [Admin API]({{< relref "../../http_api/admin.md" >}})
- [Organization API]({{< relref "../../http_api/org.md" >}})
- [Reporting API]({{< relref "../../http_api/reporting.md" >}})
- [User API]({{< relref "../../http_api/user.md" >}})
```

osg-grafana commented on docs/sources/enterprise/access-control/permissions.md:
```diff
@@ -0,0 +1,73 @@
++++
+title = "Permissions"
+description = "Understand permissions in fine-grained access control"
```
```suggestion
description = "Understand fine-grained access control permissions"
```
As odd as it is, the weight (importance) falls on the final word in this phrase.

osg-grafana commented on docs/sources/enterprise/access-control/permissions.md:
```diff
@@ -0,0 +1,73 @@
+
+# Permissions 
+
+A permission is defined by an action and a scope. When evaluating a decision for access, user will be allowed to perform a specific action on a resource if they have a permission with relevant action and scope. 
+To grant permissions to your users, you can create [built-in role assignments]({{< relref "./roles.md#built-in-role-assignments" >}}).
```
```suggestion
Each permission is defined by an action and a scope. When evaluating a fine-grain access decision, consider what specific action a user will be allowed to perform and on what resource.

To grant permissions to a user, create built-in role assignments. For more information, refer to [Built-in role assignments]({{< relref "./roles.md#built-in-role-assignments" >}}).
```

osg-grafana commented on docs/sources/enterprise/access-control/permissions.md:
```diff
@@ -0,0 +1,73 @@
+A permission is defined by an action and a scope. When evaluating a decision for access, user will be allowed to perform a specific action on a resource if they have a permission with relevant action and scope. 
+To grant permissions to your users, you can create [built-in role assignments]({{< relref "./roles.md#built-in-role-assignments" >}}).
+
+Refer to [Resources with fine-grained permissions]({{< relref "./_index.md#resources-with-fine-grained-permissions" >}}) to learn more about which permissions are used for which resources.
```
```suggestion
To learn more about what permissions are used for what resources, refer to [Resources with fine-grained permissions]({{< relref "./_index.md#resources-with-fine-grained-permissions" >}}).
```
Write "what" when considering more than two options in the context of a choice. Write "which" when the choice involves two things.

vtorosyan replied:
🚀

vtorosyan committed:
Apply suggestions from code review

Co-authored-by: Ursula Kallio <73951760+osg-grafana@users.noreply.github.com>

vtorosyan committed:
Merge branch 'main' into vt/ac-docs

vtorosyan committed:
Merge branch 'main' into vt/ac-docs

vtorosyan committed:
Better API docs and few minor fixes

mjseaman committed:
wording and description updates to _index and permissions

mjseaman committed:
add link to basic auth in HTTP API

vtorosyan committed:
More examples and fixes

vtorosyan commented:
I believe I am done adding with relevant content. From now on, hopefully there will only be edits.

mjseaman committed:
add setup/config step and link to usage scenarios

vtorosyan commented:
A friendly reminder that we are a bit tight with timeline with this PR. Beta2 goes out on Thursday, 20th of May, ideally we will have docs pushed, if not, still not the end of the world I guess, we would need to get this though before June.

osg-grafana commented:
> A friendly reminder that we are a bit tight with timeline with this PR. Beta2 goes out on Thursday, 20th of May, ideally we will have docs pushed, if not, still not the end of the world I guess, we would need to get this though before June.

Yes, and involve your friendly neighborhood technical writer early and often. Looking forward to pairing tomorrow morning.

vtorosyan committed:
Small improvements (I think)

vtorosyan committed:
Move enable section to the bottom

sakjur committed:
Access control: Remove reference to delete body

The DELETE method shouldn't use a body, so we're updating this ahead of
release.

Based on grafana/grafana-enterprise#1413

sakjur committed:
Revert "Access control: Remove reference to delete body"

This reverts commit b1dbdf9f3b1ad0bef79d2efbd6871cbfa2f7fc27.

(accidentally pushed directly to upstream branch)

vtorosyan committed:
Update docs/sources/enterprise/access-control/permissions.md

Co-authored-by: Ursula Kallio <73951760+osg-grafana@users.noreply.github.com>

vtorosyan commented:
As discussed in a group, we will be merging this PR as it is and continue working to improve the docs incrementally.

sakjur committed:
Access control: Remove reference to delete body (#34411)

* Access control: Remove reference to delete body

The DELETE method shouldn't use a body, so we're updating this ahead of
release.

Based on grafana/grafana-enterprise#1413

* Update docs/sources/http_api/access_control.md

vtorosyan committed:
Merge branch 'main' into vt/ac-docs

osg-grafana reviewed (approved):
Will be updating docs next week. Approving to unblock this for beta2.

osg-grafana reviewed (approved):
Here you go, @vtorosyan. Have a nice long weekend. You deserve it.

vtorosyan commented:
@oddlittlebird I'd appreciate if you can re-accept the PR, otherwise it is blocked for the merge. We will be working on the docs update next week, as we decided to ship the first version of docs for beta. Thanks!
